src/v2/git_url.rs

```diff
@@ -71,6 +71,44 @@ impl GitUrl {
             }
         }
     }
+
+    /// Returns a new GitUrl which is the same as the
+    /// this one, but without any subdirectory part
+    pub fn without_subdirectory(&self) -> GitUrl {
+        let (repository, branch, _) = self.parse_parts();
+        let branch_str = match branch {
+            Some(branch) => format!("#{}", branch),
+            None => String::new(),
+        };
+        GitUrl { url: format!("{}{}", repository, branch_str) }
+    }
+
+    /// Extract the repository part of the URL
+    pub fn repository(&self) -> String {
+        self.parse_parts().0
+    }
+
+    /// Extract the optional branch part of the git URL
+    pub fn branch(&self) -> Option<String> {
+        self.parse_parts().1
+    }
+
+    /// Extract the optional subdirectory part of the git URL
+    pub fn subdirectory(&self) -> Option<String> {
+        self.parse_parts().2
+    }
+
+    fn parse_parts(&self) -> (String, Option<String>, Option<String>) {
+        lazy_static! {
+            static ref URL_PARSE: Regex = Regex::new(r#"^([^#]+)(#([^:]+)?(:(.+))?)?$"#).unwrap();
+        }
+        let captures = URL_PARSE.captures(&self.url).unwrap();
```

Comment:
_This_ `unwrap`, on the other hand, looks like a possible runtime panic, because you're matching a complex regex against a user-provided string. And `unwrap`/`expect` are really meant to be assertions, not error-checks—except maybe in very short "scripts".

So we need to do one of two things here:

1. Provide an iron-clad argument that `captures` can _never_ fail on any possible input, or
2. Redesign this class / API so that we can only construct a `GitUrl` with a valid URL.

The later might look like:

 ```rust
pub struct GitUrl {
    url: String, // Must be private to maintain invariants.
} 

impl GitUrl {
    fn new<S: Into<String>>(url: S) -> Result<GitUrl> {
        let url = GitUrl { url: url.into() };
        // Make sure our URL can always be parsed.
        url.parse_parts()?;
        Ok(url)
    }
}
```

Then later you can write:

```rust
fn parse_parts(&self) -> Result<(&str, Option<&str>, Option<&str>)> { ... }

pub fn subdirectory(&self) -> Option<&str> {
    let (_, _, subdir) = self.parse_parts()
        .expect("parse_parts failed on data that we already parsed once successfully");
    subdir
}
```

The nice thing about this is that it allows us to add other validity checks to `new` in the future as needed. Alternatively, we could keep the current API, but provide a comment explaining why `parse_parts` can never fail on any possible input.

Reply:
Yeah so I was conscious of the unwrap here. I'm pretty certain that it can't blow up though. As far as I can tell, the *only* way `captures` can fail is if the url is empty, or if it starts with a hash. And either of those cases would fail the the existing `should_treat_as_url` check.

Now that said, I can totally see a situation where someone might change the `should_treat_as_url` regex, without realising that there's an `unwrap` 70 lines down that depends on it.

Conclusion: I'll go with your suggestion and make `new` call `parse_parts`, so it definitely can't panic.

Reply:
Just looking at your proposed snippet above, you've removed the `should_treat_as_url` check from `new`. I feel like we should keep that check there, seeing as it makes sure Docker will accept the URL. The `parse_parts` regex will accept strings like `"a#b:c"`, or even just `"a"`.

Reply:
Oh, yeah, definitely include whatever is already there in `new`. I wrote that `new` example from scratch.

Reply:
Ah, ok, great.

Reply:
> Now that said, I can totally see a situation where someone might change the `should_treat_as_url` regex, without realising that there's an unwrap 70 lines down that depends on it.

Yeah, a good rule of thumb is that you shouldn't call `unwrap` in a doubtful situation like this. At a minimum, there should be an `expect` and a explanation of why this really is an assertion, and not a run-time error condition.

Reply:
Ok I'm really close with this, just struggling with the final piece. Right now I have this:

```rust
    fn parse_parts(&self) -> Result<(&str, Option<&str>, Option<&str>)> {
        lazy_static! {
            static ref URL_PARSE: Regex = Regex::new(r#"^([^#]+)(?:#([^:]+)?(?::(.+))?)?$"#)
                .expect("Could not parse regex URL_PARSE");
        }
        let captures = URL_PARSE.captures(&self.url).unwrap();
        Ok((
            captures.at(1).unwrap(),
            captures.at(2),
            captures.at(3),
        ))
    }
```

I've wrapped the return type in `Result`, and I'm handling the `Result` everywhere that calls this, but as you can see I still have the `unwrap`. I tried to replace `unwrap()` with`ok_or(())?` to convert the `Option` into a `Result` but couldn't get the types to work. Is it because the `Result` here is an error-chain thing, not a regular `Result`?

Reply:
Yeah, you need to construct our local Error type here, which you can actually do with a string and .into(). Use `ok_or_else(|| -> Error { format!(...).into() })` and it will construct an error for you.